### ISSUE TITLE
Permission denied in the /etc/openvpn/tmp folder when plugin script is called

### ISSUE THREAD

**Issue by deji** [okta/okta-openvpn#14]
I setup my OpenVPN server to switch to user and group openvpn after initialisation. This appears to lead to a circular situation where okta-openvpn plugin script fails with ControlFilePermissionsError in line 394 of okta_openvpn.py when attempting to write to the tmp folder at /etc/openvpn/tmp.

When the mode is set to 0700 and the owner is openvpn, the service is unable to start. When the owner is switched to root, the service starts but the plugin fails. Giving write permission to the group also causes the plugin script to fail.

How can I get around this problem with leaving the OpenVPN server running as root?

Thanks.

**Comment by zpol** [comment]
@deji  did you solve this? I have the same issue. 

Thanks in advance

**Comment by deji** [comment]
I went the route of setting uid/gid to root :(




regards,
Deji.

sent from my oneplus3

**Comment by deji** [comment]
I'm revisiting this again @zpol and I still can't get it to work without uid/gid set to root. How did you get yours to work? What uid/gid are you using?

Thanks.

**Comment by rduboiscortex** [comment]
I'm running into this issue as well. The only way I found to fix the issue is to switch the owner of the tmp folder from root to nobody after the OpenVPN has been initialized and no users have logged in.

Is is possible to run this plugin as a non-privileged user ?

**Comment by zpol** [comment]
Hi everybody, I just changed the permissions and it started working. But I don't have access to that instance anymore, it was an AWS instance so I destroyed it be¡cause after several tests, finally I switched to Pritunl,  a very improoved vpn software based on openvpn source code and with many cool features. It's also connected with Okta via SAML for SSO and it can be managed from a very nice web GUI. Everything can be installed & config even with Okta in minutes.

**Comment by joelfranusic-okta** [comment]
Here is the part of the source code that is causing the issues: https://github.com/okta/okta-openvpn/blob/master/okta_openvpn.py#L378-L396

In short, you should:
1. Have a special folder just for OpenVPN, I suggest `/etc/openvpn/tmp`
2. Make sure the folder is owned by the same user that OpenVPN is run as
3. Make sure that nobody else can write to that folder

Here is where this /should/ be documented: https://github.com/okta/okta-openvpn/#make-sure-that-openvpn-has-a-tempory-directory - I'd appreciate a pull request to the docs to help make them more clear!

**Comment by chad-cook** [comment]
So i've spent the day on this. I can see where the problem is. 

TLDR: This will only work if you are running OpenVpn as root.

Explanation: 

- In the server.conf there is an option for "user" and "group" to harden OpenVpn most people set this to nobody and nogroup or a service account.
- OpenVpn starts as root and then changes down to the user stipulated in the conf file.
- When OpenVpn starts it checks that it has the correct permissions to its "tmp-dir" it then changes to down to the "user" provided.
-  The code provided here checks that only the OpenVpn service has access to the folder.

Recreate:
- Create a service user "openvpn"
- Create folder "tmp"
- Chown openvpn:root "tmp"
- Chmod 700 "tmp"
- Add line server.conf "user openvpn"
- Add line server.conf "tmp-dir "/etc/openvpn/tmp"
- Start OpenVPN and you will get an error message that openvpn can't access that folder.

ok so we change permssions so root has access as well.
- Chmod 770 "tmp"
- Start OpenVPN (It works yay)
- we authenticate and the python code denies due to a permission issue.

Solution:
Basically the service needs to be run as root unless we do the following;
- chmod 770 tmp so both root and the openvpn user have access
- remove the check for permissions

**Comment by joelfranusic-okta** [comment]
Hey @chad-cook – thanks so much for this in depth writeup. This is the second time that this issue has come up in a week. Clearly I need to address this. 

What would you suggest as the best way to solve this issue for your use case?

**Comment by chad-cook** [comment]
That's a tough question. It's a bit of a rock and a hard place. I pressed forward with dropping the call to the check permissions function. 

In saying that it was late and now something else is broken. 

Thank you for picking this back up

**Comment by diepes** [comment]
why does openvpn need a tmp folder/file ?
